frontend/controller/actions/contract.js

```diff
@@ -0,0 +1,23 @@
+import sbp from '~/shared/sbp.js'
+import { CONTRACT_IS_SYNCING } from '@utils/events.js'
+
+export default sbp('sbp/selectors/register', {
+  'gi.actions/contract/subscribe': async function (contractIDs: string | array) {
+    const listOfIds = typeof contractIDs === 'string' ? [contractIDs] : contractIDs
+
+    for (const id of listOfIds) {
+      await sbp('state/enqueueContractSync', id)
+    }
+  },
+  'gi.actions/contract/subscribeAndWait': async function (contractIDs: string | array) {
```

Comment:
@taoeffect, regarding this domain API, you suggested:

> - gi.actions/contract/subscribe <-- takes a string or an array of strings (representing contractIDs) and calls sbp('state/enqueueContractSync', <hash>) on them
> - gi.actions/contract/subscribeAndWait <-- takes a string or an array of strings (representing contractIDs), creates an array of promises that each register an event listener for CONTRACT_IS_SYNCING on for their contractID, and resolve when the event listener is called with isSyncing = false (see syncContractWithServer function in state.js). Then calls gi.actions/contract/subscribe on the string (or array of strings), and then returns Promise.all(arrayOfPromises).

The same for the user and group domains. Build a `/create|subscribe` method and 1 similar method `andWait`. My question is, why not creating only one method that by default waits for CONTRACT_IS_SYNCING?

Reply:
1. That's a good suggestion. In fact, I think the way `'gi.actions/contract/sync'` works might already do this! See my comment below.
2. Sure, I've provided the implementation I had in mind below. :)